### PR TITLE
Add fuzzer for OpenAPI V3

### PR DESCRIPTION
**Pull request by Jefftree**
Marking as WIP because it may still be incomplete, most of the important things should be there though.

/cc @apelisse @alexzielenski

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,13 +6,15 @@ require (
 	github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46
 	github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
 	github.com/emicklei/go-restful/v3 v3.8.0
+	github.com/getkin/kin-openapi v0.76.0
```
Comment (Jefftree) [inline go.mod:9]: I recall that we did a fair bit of work to ensure that kin-openapi isn't getting imported in k/k. Should the fuzzer testing be moved into integration instead?

File: pkg/spec3/fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package spec3
+
+import (
+	"encoding/json"
+
+	"testing"
+
+	"github.com/getkin/kin-openapi/openapi3"
```
Comment (apelisse) [inline pkg/spec3/fuzz_test.go:8]: I could drop this test and get rid of the dependency. Why do you work so hard trying to make sure that the fuzzer generates entirely valid openapi?

**Comment by apelisse** [comment]
I'd love @alexzielenski to make a first pass. He's a lot more familiar with this than I am.

File: pkg/spec3/fuzz_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package spec3
```
Comment (alexzielenski) [inline pkg/spec3/fuzz_test.go:1]: can this be blackbox test?

**Review by alexzielenski — CHANGES_REQUESTED** [review]

File: pkg/spec3/response.go [hunk 4]
```diff
@@ -78,20 +78,29 @@ func (r ResponsesProps) MarshalJSON() ([]byte, error) {
 
 // UnmarshalJSON unmarshals responses from JSON
 func (r *ResponsesProps) UnmarshalJSON(data []byte) error {
-	var res map[string]*Response
+	var res map[string]json.RawMessage
```
Comment (Jefftree) [inline pkg/spec3/response.go:81]: This change will be removed here and included in a separate PR: https://github.com/kubernetes/kube-openapi/pull/372

File: pkg/spec3/benchmark_serialization_test.go [hunk 5]
```diff
@@ -19,6 +20,29 @@ var swaggerDiffOptions = []cmp.Option{cmp.Comparer(func(a spec.Ref, b spec.Ref)
...
+	}
+	var actual *OpenAPI
+	err = json.Unmarshal(j, &actual)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(expected, actual) {
+		t.Fatal(cmp.Diff(expected, actual, swaggerDiffOptions...))
+	}
+}
```
Comment (apelisse) [inline pkg/spec3/benchmark_serialization_test.go:44]: Should we also check that marshaling is stable too?

File: pkg/spec3/fuzz.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
+package spec3
+
+import (
+	fuzz "github.com/google/gofuzz"
+	"github.com/lucasjones/reggen"
```
Comment (apelisse) [inline pkg/spec3/fuzz.go:5]: Can we add this dependency? That seems unfortunate. Any other way to generate valid arbitrary strings?

Comment (Jefftree) [reply pkg/spec3/fuzz.go:5]: It's a single go file https://github.com/lucasjones/reggen/blob/master/reggen.go, if it's not a desired dependency, I can write a custom function to generate valid arbitrary strings.

File: pkg/spec3/benchmark_serialization_test.go (hunk 5 above)
Comment (Jefftree) [reply pkg/spec3/benchmark_serialization_test.go:44]: Isn't that something that's given based on the conditions here?

We don't have a OpenAPI JSON fuzzer (and I don't think we should build one) so the base case starts from a fuzzed OpenAPI that's marshalled. The unmarshal test above proves x == Unmarshal(Marshal(x)) where x is a fuzzed OpenAPI. Writing a new test to for Unmarshal is trying to determine the relationship between Marshal(x) and Marshal(Unmarshal(Marshal(x))). Since we already showed x == Unmarshal(Marshal(x)), the latter simplifies to Marshal(x). The two would only not equal if Marshal is not stable and wouldn't exactly be a roundtrip test? 

When we implement the experimental marshaler, then it makes sense to add serialization tests to ensure that the old and new marshaler have the same behavior. WDYT?

File: pkg/spec3/benchmark_serialization_test.go [hunk 7]
```diff
@@ -19,6 +20,29 @@ var swaggerDiffOptions = []cmp.Option{cmp.Comparer(func(a spec.Ref, b spec.Ref)
...
+	var fuzzer *fuzz.Fuzzer
+	fuzzer = fuzz.NewWithSeed(1646791953)
+	// Make sure we have enough depth such that maps do not yield nil elements
+	fuzzer.MaxDepth(22).NilChance(0.5).NumElements(1, 7)
+	fuzzer.Funcs(OpenAPIV3FuzzFuncs...)
+	expected := &OpenAPI{}
+	fuzzer.Fuzz(expected)
+
+	j, err := json.Marshal(expected)
+	if err != nil {
```
Comment (alexzielenski) [inline pkg/spec3/benchmark_serialization_test.go:33]: I turned down the nil chance, and ran this test in a loop with a few random seeds, and found something is up with `Paths` that doe snot round trip properly

**Review by alexzielenski — CHANGES_REQUESTED** [review]

File: pkg/spec3/benchmark_serialization_test.go [hunk 8]
```diff
@@ -19,6 +20,29 @@ var swaggerDiffOptions = []cmp.Option{cmp.Comparer(func(a spec.Ref, b spec.Ref)
 	return a.String() == b.String()
 })}
 
+func TestOpenAPIV3RoundTrip(t *testing.T) {
```
Comment (alexzielenski) [inline pkg/spec3/benchmark_serialization_test.go:23]: https://github.com/kubernetes/kube-openapi/blob/f73e7dbcfe2954706d8ee5db7f1adaed3950e8de/pkg/util/jsontesting/json_roundtrip.go#L28-L44

might be suitable to use here

Comment (Jefftree) [reply pkg/spec3/benchmark_serialization_test.go:23]: To conform to the interface, Unmarshal is implemented in https://github.com/kubernetes/kube-openapi/pull/368, can I fix this in a follow up?

File: pkg/spec3/benchmark_serialization_test.go (hunk 7 above)
Comment (Jefftree) [reply pkg/spec3/benchmark_serialization_test.go:33]: Fixed. Accidentally removed the requirement that paths needed to start with a slash in a refactor.

**Review by alexzielenski — CHANGES_REQUESTED** [review]
only saw small thing, then lgtm

File: pkg/spec3/fuzz.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+			return
+		}
+		c.Fuzz(&v.VendorExtensible)
+		c.Fuzz(&v.Description)
+		c.Fuzz(&v.Nullable)
+		c.Fuzz(&v.Title)
+		c.Fuzz(&v.Required)
+		c.Fuzz(&v.ExternalDocs)
+		n := c.Intn(8)
+		switch n {
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/spec3/fuzz.go:197]: That makes sense, I did not realize.

File: pkg/spec3/fuzz.go [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+		}
+		for i := 0; i < numChildren; i++ {
+			v.Add("x-"+randAlphanumString(), c.RandString()+"x")
+		}
+	},
+	func(v *spec.ExternalDocumentation, c fuzz.Continue) {
+		c.Fuzz(&v.Description)
+		v.URL = "https://" + randAlphanumString()
+	},
+	func(v *spec.Schema, c fuzz.Continue) {
```
Comment (apelisse) [inline pkg/spec3/fuzz.go:185]: I have only checked quickly but this is missing a lot of fields, which means that if these fields are not round-tripped properly, we're very likely to miss that in the tests.

Comment (Jefftree) [reply pkg/spec3/fuzz.go:185]: Let me see if we can reuse the fuzzer that Alex has already wrote for this.
